Refactor `getPermissionsAsync` on expo-location to query for user permission instead of request permission

hems opened the pull request:
# Why

See this issue https://github.com/expo/expo/issues/26836


# How

Using new navigator.permissions API

# Test Plan

I could not find a test for the web platform in this package, also i didn't have enough time to dig for other web tests in the repo.

If someone could help me out write the test here and/or point me out to other similar tests ( web platform with mocks, etcs ) i would happily have a look at it.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- ^ documentation doesn't seem to mention this **"unexpected behaviour"**
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- ^ i haven't changed the documentation
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- ^ how can i confirm this? i didn't change anything significant, so should still be OK.

hems committed:
updating expo-location to use permissions api on web platform

lukmccall commented on packages/expo-location/src/ExpoLocation.web.ts:
```diff
@@ -49,30 +49,33 @@ function isLocationValid(location: LocationObject, options: LocationLastKnownOpt
+ * possible to query for permission on all browsers, apparently only the
+ * latest versions will support this.
  */
 async function getPermissionsAsync(): Promise<PermissionResponse> {
```
```suggestion
async function getPermissionsAsync(): Promise<PermissionResponse> {
  if (!navigator?.permissions?.query) {
    throw new UnavailabilityError(
      'expo-location',
      'navigator.permissions API is not available'
    );
  }
```
Let's check if the API is available.

lukmccall commented on packages/expo-location/src/ExpoLocation.web.ts:
```diff
@@ -49,30 +49,33 @@ function isLocationValid(location: LocationObject, options: LocationLastKnownOpt
-      },
-      { enableHighAccuracy: false, maximumAge: Infinity }
-    );
+    navigator.permissions.query({name: 'geolocation'}).then((permission) => {
```
Can we use `await` instead of callbacks? 
```js
const { state } = await navigator.permissions.query({ name: 'geolocation' });
```

hems replied:
defo!

lukmccall commented on packages/expo-location/src/ExpoLocation.web.ts:
```diff
@@ -49,30 +49,33 @@ function isLocationValid(location: LocationObject, options: LocationLastKnownOpt
+ * latest versions will support this.
  */
 async function getPermissionsAsync(): Promise<PermissionResponse> {
   return new Promise<PermissionResponse>((resolve) => {
```
We can return a result when using the `async` syntax instead of callbacks.

lukmccall commented on packages/expo-location/src/ExpoLocation.web.ts:
```diff
@@ -49,30 +49,33 @@ function isLocationValid(location: LocationObject, options: LocationLastKnownOpt
-      { enableHighAccuracy: false, maximumAge: Infinity }
-    );
+    navigator.permissions.query({name: 'geolocation'}).then((permission) => {
+      if(permission.state === 'granted') {
+        resolveWithStatus(PermissionStatus.GRANTED)
+      } else if(permission.state === 'denied') {
+        resolveWithStatus(PermissionStatus.DENIED, false)
+      } else if(permission.state === 'prompt') {
+        // on Chrome, the permission state is 'prompt' when the permission has
+        // not been requested yet
+        resolveWithStatus(PermissionStatus.UNDETERMINED)
+      } else {
+        resolveWithStatus(PermissionStatus.UNDETERMINED)
+      }
```
```suggestion
      if (permission.state === 'granted') {
        resolveWithStatus(PermissionStatus.GRANTED)
      } else if (permission.state === 'denied') {
        resolveWithStatus(PermissionStatus.DENIED, false)
      } else if (permission.state === 'prompt') {
        // on Chrome, the permission state is 'prompt' when the permission has
        // not been requested yet
        resolveWithStatus(PermissionStatus.UNDETERMINED)
      } else {
        resolveWithStatus(PermissionStatus.UNDETERMINED)
      }
```

lukmccall reviewed (commented):
It looks good. We can use the new permission API. I've left some comments. Can you also add a changelog entry?

hems committed:
adding entry to changelog, fixing typo on log

hems committed:
checking if navigator permission exists, using async/await instead of callback

hems commented:
Thanks for the review @lukmccall i have incorporated all your suggestions and i have also added an entry to the CHANGELOG.

I wonder what would be the procedure to "compile" ( or build ) this package and test it out?

Is there any package for `web` that includes tests that would run on a "test browser" like [playwright](https://playwright.dev/) ?

hems commented:
@lukmccall / @byCedric let me know if i'm missing something here in order to have this merged.

thanks a lot.

lukmccall commented:
Hey @hems 👋  Sorry for the late reply. Can you rebuild the js files in the `expo-location` by running `yarn build` in the `expo-location` package? After that, I think we can merge it.

hems committed:
running yarn build on expo-location folder

hems commented:
> Hey @hems 👋 Sorry for the late reply. 

no problems at all. i have been AFK last weeks as well, happy to pick this back and get it merged!

> Can you rebuild the js files in the `expo-location` by running `yarn build` in the `expo-location` package? 

I just ran build and executed "git diff" afterwards in order to see what changed but for my surprise git claimed the files were binary and didn't show me the difference. Is that something expected? 

![image](https://github.com/expo/expo/assets/27327/a5243835-c23f-4e6d-b090-9664f1641981)

> After that, I think we can merge it.

It's now pushed into this PR.

One of my concerns is that this could potentially break web applications that were relying on his unexpected/unwanted behaviour, in other words, this "update" / "fix" / "improvement" might introduce breaking changes to the API to the ( very likely ) small amount of users ( who like me ) might be relying on this library to build web pages

lukmccall commented:
I've rebased your code in https://github.com/expo/expo/pull/27724. 

> One of my concerns is that this could potentially break web applications that were relying on his unexpected/unwanted behaviour, in other words, this "update" / "fix" / "improvement" might introduce breaking changes to the API to the ( very likely ) small amount of users ( who like me ) might be relying on this library to build universal apps and web pages with the same code base.

I was worried about an issue and took action to address it. I modified the logic slightly by modifying the request permissions method. Now, when the permissions state is `prompt`, the method calls `getCurrentLocation` to trigger the prompt. In users' app, the regular flow should look like this:
```
const { granted } = await getPermissions()
if (!granted) {
  const response = await requestPermissions();
  // do something with response 
}
```
If a user has code like this, the change won't break anything, thereby minimizing the affected apps. 

I'll merge the new version of that PR. Thanks for your contribution!
